### ISSUE TITLE
TOOL Security: Update formflow.ai to Latest Version 4.0.5

### ISSUE THREAD

**Issue by warrenchristian1telus** [bcgov/digital-journeys#275]
Update composer dependencies in the repository to eliminate any unnecessary security vulnerabilities.

This is in relation to Igor's report, and Dependabot security Alerts:

https://github.com/bcgov/digital-journeys/security/dependabot

I think we have two main options to resolve at least the majority of these issues. Some vulnerabilities likely do not yet have a fix available, so may need to be mitigated in other ways.

1. Delete package-lock.json from the repository
      a) This is my preference, and what we have done for the Moodle project
      b) Every published code change (or automated build) will always download the latest (minor) updates
          i) This allows for automated patches on a regular basis
          ii) Introduces risk that a code change could have greater / unexpected impacts in other areas
              1) Automated functional testing is recommended to reduce this risk
2. Have developers update composer files locally, test / validate functionality, then deploy the updated package-lock.json via Git to the repo
       a) This would require a GitHub issue to be created (for each update / release)
       b) Could be triggered by alerts (Dependabot, etc.)
       c) Could be set based on time / resourcing
             i) Once per sprint
             ii) Every code change / deployment
             iii) Weekly, monthly, etc.
        d) This allows for the most control, but also takes the most time / resources
        e) The risk with this approach is that once active development slows or ceases, updates may be neglected

I think the logical first step is to delete or update the package-lock.json file to see how many vulnerabilities we can clear up with a quick update / deployment.

Either option is very easy to run in dev and revert if there are any issues.

**Comment by warrenchristian1telus** [comment]
I am discussing options / best practices for handling security alerts with the other developers / teams.

Currently the top pics for scanning our code and deploying (automatic) code updates where possible are Snyk and Dependabot. Both have been installed / enabled on our repository, and are successfully scanning and identifying vulnerabilities.

We will need to compare these results with those from Aqua (which scans our deployed containers in OpenShift) to determine which one(s) we should be using as a primary resource for identifying and deploying patches.

We will also need to create a policy around how and when to deploy automated changes, and when and whom should determine if/when/who should investigate and remediate issues that don't have an automated and/or published solution.

**Comment by warrenchristian1telus** [comment]
An example of a fix that requires additional investigation is this recommendation from Snyk to change the base image of forms-flow-api from python:3.8 to python:3.8-slim-bullseye. The slim-bullseye version has less vulnerabilities, but we would need to determine / test if it is a viable option for our particular environment and configuration.

https://app.snyk.io/org/warrenchristian1telus/project/dbc02ce5-3d17-493f-a275-46654b9a0429

![image.png](https://images.zenhubusercontent.com/5f6b6e00c344a38c20855fad/34c3e296-8fc9-458d-89a8-cb0ad863e0b3)

**Comment by warrenchristian1telus** [comment]
Another thing I would like to mention is that there is no known exploit for any of these vulnerabilities, and that KeyCloak provides a significant level of protection for all of our internet-facing applications. I think managing threat assessments and overall risk should be a primary consideration for how we deal with these issues.

Determining which issues we fix, and when, can make a big difference to developer resourcing and project development timelines. Of course we don't want to leave our applications vulnerable to attack, but we should also be able to assess which ones we can safely / reasonably ignore or postpone.

**Comment by warrenchristian1telus** [comment]
I've upgraded python from 3.8 to 3.8-slim-bullseye to reduce vulnerabilities by reducing the default packages installed. Deployment was successful. We're now reviewing and testing for any issues that may have been introduced in dev.

**Comment by warrenchristian1telus** [comment]
Upon closer inspection, although this addresses 35 vulnerabilities, they are all in one file / server (forms-flow-bpm). 

https://github.com/bcgov/digital-journeys/pull/365/commits/9b5b9cd6388768d9a514084a0872a0a7550897db

This is only one of 14 deployments / servers that will need to be addressed, and only 5 of which can be (/have been) scanned by Advanced Cluster Security.

Dependabot has identified 63 additional vulnerabilities: https://github.com/bcgov/digital-journeys/security/dependabot

I can enable more version auto-updates from Dependabot, but that will require a code change to the repo. If we want to do that, we should have a new issue to address it. I've created one here so we can address it if/when we're ready to deploy: https://app.zenhub.com/workspaces/digital-journeys-61c10fe84c8bcb001491c5f5/issues/bcgov/digital-journeys/366

I have also disabled the deployment pipeline for now, to allow for a few code updates (as above) and Pull Requests to be merged into the repo without each triggering a deployment. This way, we can release (and roll back) everything at once.

This may also be a good time to verify our backups / process. As this isn't a single PR anymore, I think it may be good to involve @kkapoor-fresh and try to ensure that we will have wide update coverage and a solid rollback strategy. With so many code changes, in so many files, I think we should be ready to test our disaster recovery strategy (for rebuilding dev).

To accommodate this, I'll also assign Kunal and increase the effort estimate to 5. I'm hoping this will still be fairly easy to implement, but want to be a bit more cautious and create some additional documentation as the process is getting more complex.
